<h1>Requests to Books subscription service databases<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Tale-and-aim" data-toc-modified-id="Tale-and-aim-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Tale and aim</a></span></li><li><span><a href="#Specified-requests" data-toc-modified-id="Specified-requests-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Specified requests</a></span></li><li><span><a href="#Importing-libraries" data-toc-modified-id="Importing-libraries-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Importing libraries</a></span></li><li><span><a href="#EDA" data-toc-modified-id="EDA-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>EDA</a></span><ul class="toc-item"><li><span><a href="#books" data-toc-modified-id="books-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span><code>books</code></a></span></li><li><span><a href="#authors" data-toc-modified-id="authors-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span><code>authors</code></a></span></li><li><span><a href="#publishers" data-toc-modified-id="publishers-4.3"><span class="toc-item-num">4.3&nbsp;&nbsp;</span><code>publishers</code></a></span></li><li><span><a href="#ratings" data-toc-modified-id="ratings-4.4"><span class="toc-item-num">4.4&nbsp;&nbsp;</span><code>ratings</code></a></span></li><li><span><a href="#reviews" data-toc-modified-id="reviews-4.5"><span class="toc-item-num">4.5&nbsp;&nbsp;</span><code>reviews</code></a></span></li><li><span><a href="#Conclusions-on-EDA" data-toc-modified-id="Conclusions-on-EDA-4.6"><span class="toc-item-num">4.6&nbsp;&nbsp;</span>Conclusions on EDA</a></span></li></ul></li><li><span><a href="#Performing-the-specified-requests" data-toc-modified-id="Performing-the-specified-requests-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Performing the specified requests</a></span><ul class="toc-item"><li><span><a href="#Number-of-books,-published-after-January-the-1st,-2000" data-toc-modified-id="Number-of-books,-published-after-January-the-1st,-2000-5.1"><span class="toc-item-num">5.1&nbsp;&nbsp;</span>Number of books, published after January the 1st, 2000</a></span></li><li><span><a href="#Number-of-user-reviews-and-average-user-rating-for-every-book" data-toc-modified-id="Number-of-user-reviews-and-average-user-rating-for-every-book-5.2"><span class="toc-item-num">5.2&nbsp;&nbsp;</span>Number of user reviews and average user rating for every book</a></span></li><li><span><a href="#A-publisher,-which-has-published-the-greatest-number-of-books-containing-more-than-50-pages" data-toc-modified-id="A-publisher,-which-has-published-the-greatest-number-of-books-containing-more-than-50-pages-5.3"><span class="toc-item-num">5.3&nbsp;&nbsp;</span>A publisher, which has published the greatest number of books containing more than 50 pages</a></span></li><li><span><a href="#An-author-who-got-the-highest-average-user-rating---to-consider-only-the-books-with-50-and-more-ratings)" data-toc-modified-id="An-author-who-got-the-highest-average-user-rating---to-consider-only-the-books-with-50-and-more-ratings)-5.4"><span class="toc-item-num">5.4&nbsp;&nbsp;</span>An author who got the highest average user rating - to consider only the books with 50 and more ratings)</a></span></li><li><span><a href="#An-average-number-of-reviews-from-users-who-gave-more-than-50-ratings" data-toc-modified-id="An-average-number-of-reviews-from-users-who-gave-more-than-50-ratings-5.5"><span class="toc-item-num">5.5&nbsp;&nbsp;</span>An average number of reviews from users who gave more than 50 ratings</a></span></li></ul></li><li><span><a href="#Conclusions" data-toc-modified-id="Conclusions-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Conclusions</a></span></li></ul></div>

Start <a class="anchor" id="100-bullet"></a>

## Tale and aim

The coronavirus took the world off guard, changing the usual order of things. In their free time city dwellers no longer go out, do not visit cafes and shopping centers. But there is more time for books. Startups noticed this - and rushed to create applications for those who like to read.
A company decided to be on the wave and bought a major subscription book reading service. 
The job now is to analyze its database. The datasets contain information about books, publishers, authors, user reviews and user ratings of books. Getting data by specified requests should help the company to formulate a value proposition for service's promotion.

## Specified requests

1. To count number of books, published after January the 1st, 2000;
2. To count number of user reviews and average user rating for every book;
3. To determine a publisher, which has published the greatest number of books containing more than 50 pages (to exclude brochures from analysis);
4. To determine an author who got the highest average user rating - to consider only the books with 50 and more ratings;
5. To calculate an average number of reviews from users who gave more than 50 ratings.


NOTA BENE: The term of the tasks - to get requied data only by means of SQL and to perform only one SQL-request per task. Pandas is allowed only for data preprocessing, displaying and storing results.

## Importing libraries

In [1]:
import pandas as pd
from sqlalchemy import create_engine

In [2]:
# setting parameters
db_config = {'user': 'praktikum_student', # login
'pwd': 'Sdf4$2;d-d30pp', # password
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # port of connection
'db': 'data-analyst-final-project-db'} # database's name
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(
db_config['user'],
db_config['pwd'],
db_config['host'],
db_config['port'],
db_config['db'])

# saving the connector
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

[Come back to Start](#100-bullet)

## EDA

In [34]:
# make a function for reading sql
def read_sql(query):
    result = pd.io.sql.read_sql(query, con = engine)
    return result

Let's look into the first rows of the tables and consider if any preprocessing's needed.

### `books`

In [35]:
# download first 5 rows from `books` dataset
query_1 = '''
select *
from books
limit 5;
'''
read_sql(query_1)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

Words in column `title` are written in different registers - this might cause duplicates.

In [36]:
# how big is `books` dataset
query_2 = '''
select count (*) 
from books;
'''
read_sql(query_2)

count
0   1000

The dataset `books` contains 1000 rows and 6 columns:
- `book_id` — identifier of a book;
- `author_id` — identifier of an author;
- `title` — book's title;
- `num_pages` — number of pages;
- `publication_date` — date of book's publication;
- `publisher_id` — identifier of a publisher.

Let's check duplicates through counting unique `book_id` and unique `title` (make request for lower register), as well count unique `author_id` and unique `publisher_id`.

In [47]:
query_3 = '''
select count (distinct book_id) as book_id_uniq_cnt
    , count (distinct lower (title)) as title_uniq_cnt
    , count (distinct author_id) as author_id_uniq_cnt
    , count (distinct publisher_id) as publisher_id_uniq_cnt
from books;
            '''
read_sql(query_3)

book_id_uniq_cnt  title_uniq_cnt  author_id_uniq_cnt  publisher_id_uniq_cnt
0              1000             999                 636                    340

The dataset `books`contains info on 1000 books written by 636 authors and published by 340 publishers. Seems, two books in the dataset have a same title. Let's check if the books with same title are duplicates or various publications.

In [71]:
# having selected only repepeating books' titles in the subquery we take out from the `books` dataset all data on these 
# books' titles
query_4 = '''
select *
from books
where lower (title) in (
select sub.title
from (
select lower (title) as title
    , count (lower (title)) as title_cnt
from books
group by lower (title)
having count (lower (title)) > 1) as sub
);
                        '''
read_sql(query_4)

book_id  author_id                title  num_pages publication_date  \
0      426         39  Memoirs of a Geisha        434       2005-11-15   
1      427         39  Memoirs of a Geisha        503       2005-11-22   

   publisher_id  
0           241  
1           311

Yes, the dataset `books` contains two publications of the book "Memoirs of a Geisha", performed by two publishers -  these are not duplicates.

[Come back to Start](#100-bullet)

### `authors`

In [10]:
# download first 5 rows from `authors` dataset
query_5 = ''' 
select * 
from authors 
limit 5;
'''
read_sql(query_5)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

The same thing with different registers in `author` column.

In [72]:
# how big is `authors` dataset
query_6 = '''
select count (*) 
from authors;
'''
read_sql(query_6)

count
0    636

The dataset `authors` contains 636 rows (coincides with number of unique `author_id` in `books`-dataset) and 2 columns:
- `author_id` — identifier of an author;
- `author` — author's name.

In [73]:
# let's count unique authors' names, using `lower` for register-issue's neutralization
query_7 = '''
select count (distinct author_id) as author_id_uniq_cnt
    , count (distinct lower (author)) as author_uniq_cnt
from authors;
            '''
read_sql(query_7)

author_id_uniq_cnt  author_uniq_cnt
0                 636              636

There're no duplicates revealed in the table `books`.

[Come back to Start](#100-bullet)

### `publishers`

In [74]:
# download first 5 rows from `publishers` dataset
query_8 = ''' 
select * 
from publishers 
limit 5;
'''
read_sql(query_8)

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [83]:
# check how big is `publishers` dataset and count unique publishers' names, using `lower` for register-issue's neutralization
query_9 = '''
select count (*) as rows_cnt
    , count (distinct publisher_id) as publisher_id_uniq_cnt
    , count (distinct lower (publisher)) as publisher_uniq_cnt
from publishers;
'''
read_sql(query_9)

count  publisher_id_uniq_cnt  publisher_uniq_cnt
0    340                    340                 339

The dataset `publishers` contains 340 rows (coincides with number of unique `publisher_id` in `books`dataset) and 2 columns:
- `publisher_id` — identifier of a publisher;
- `publisher` — publisher's name.

And, it seems, there's one duplicate.

In [80]:
# having selected only repepeating publishers' names in the subquery we take out from the `publishers`dataset all data
# on these repepeating publishers
query_11 = '''
select *
from publishers
where lower (publisher) in (
select sub.publisher
from (
select lower (publisher) as publisher
    , count (lower (publisher)) as publisher_cnt
from publishers
group by lower (publisher)
having count (lower (publisher)) > 1) as sub
);
                        '''
read_sql(query_11)

publisher_id         publisher
0            30  BALLANTINE BOOKS
1            33  Ballantine Books

Let's be aware that we have duplicates for the publisher "Ballantine Books": one time the publisher is listed under  `publisher_id` = 30, another time under `publisher_id` = 33. And this is not the publisher which published any of  "Memoirs of a Geisha". Another question about "Ace book" and "Ace Books", is it a typo or there're such different publishers.

[Come back to Start](#100-bullet)

### `ratings`

In [81]:
# download first 5 rows from `publishers` dataset
query_12 = ''' 
select * 
from ratings 
limit 5;
'''

read_sql(query_12)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [90]:
query_13 = '''
select count (*) as rows_cnt
    , count (distinct rating_id) as rating_id_uniq_cnt
    , count (distinct book_id) as book_id_uniq_cnt
    , count (distinct lower (username)) as username_uniq_cnt
from ratings;
'''
read_sql(query_13)

rows_cnt  rating_id_uniq_cnt  book_id_uniq_cnt  username_uniq_cnt
0      6456                6456              1000                160

The dataset `ratings` contains 6456 rows (coincides with number of unique `rating_id`) and 4 columns:
- `rating_id` — identifier of a rating;
- `book_id` — identifier of a book;
- `username` — name of a user giving the rating;
- `rating` — rating for a book.

Seems, all is ok with register. There're 160 unique users, who gave rating, and all 1000 books have got ratings.

[Come back to Start](#100-bullet)

### `reviews`

In [87]:
# download first 5 rows from `reviews`  dataset
query_14 = ''' 
select *
from reviews 
limit 5;
'''
read_sql(query_14)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

In [92]:
query_15 = '''
select count (*) as rows_cnt
    , count (distinct review_id) as review_id_uniq_cnt
    , count (distinct book_id) as book_id_uniq_cnt
    , count (distinct lower (username)) as username_uniq_cnt
    , count (distinct lower (text)) as text_uniq_cnt
from reviews;
'''
read_sql(query_15)

rows_cnt  review_id_uniq_cnt  book_id_uniq_cnt  username_uniq_cnt  \
0      2793                2793               994                160   

   text_uniq_cnt  
0           2793

The dataset `reviews` contains 2793 rows (coincides with number of unique `review_id`) and 4 columns:
- `review_id` — identifier of a review;
- `book_id` — identifier of a book;
- `username` — name of a user giving the review;
- `text` — review's text.

Seems, all is ok with register. There're 160 unique users, who wrote reviews, and but not all 1000 books have got them - reviews were given for 994 books. There're no duplicates revealed.

[Come back to Start](#100-bullet)

### Conclusions on EDA

The dataset `books`contains info on 1000 books written by 636 authors and published by 340 publishers. There're two publications of the book "Memoirs of a Geisha", performed by two publishers, duplicates haven't been revealed.

The dataset `authors` contains 636 rows, duplicates haven't been revealed.

The dataset `publishers` contains 340 rows. There're revealed duplicates for the publisher "Ballantine Books": one time the publisher is listed under `publisher_id` = 30, another time under `publisher_id` = 33. And this is not the publisher which published any of  "Memoirs of a Geisha". Another question about "Ace book" and "Ace Books", is it a typo or there're such different publishers.

The dataset `ratings` contains 6456 rows. There're 160 unique users, who gave rating, and all 1000 books have got ratings.

The dataset `reviews` contains 2793 rows. There're 160 unique users, who wrote reviews, and but not all 1000 books have got them - reviews were given for 994 books. There're no duplicates revealed.

[Come back to Start](#100-bullet)

## Performing the specified requests

### Number of books, published after January the 1st, 2000

In [94]:
query_16 = '''
select 
    count(book_id) as cnt
from books
where publication_date > '2000-01-01';
'''
read_sql(query_16)

cnt
0  819

[Come back to Start](#100-bullet)

### Number of user reviews and average user rating for every book

Viariant 1

In [105]:
query_17_1 = '''
select
    books.book_id,
    books.title,
    sub1.review_count,
    sub2.avg_rating

from books
    left join
        (select 
            book_id,
            count (review_id) as review_count
        from reviews
        group by book_id) as sub1 on books.book_id = sub1.book_id
    
    left join 
        (select 
            book_id,
            avg (rating) as avg_rating
        from ratings
        group by book_id) as sub2 on books.book_id = sub2.book_id
order by avg_rating;
'''
read_sql(query_17_1)

book_id                                              title  review_count  \
0        303                               Harvesting the Heart           2.0   
1        202                                      Drowning Ruth           3.0   
2        371                                              Junky           2.0   
3        316                  His Excellency: George Washington           2.0   
4        915  The World Is Flat: A Brief History of the Twen...           3.0   
..       ...                                                ...           ...   
995      182                                         Dead Souls           2.0   
996      599  Stone of Farewell (Memory  Sorrow  and Thorn  #2)           1.0   
997       62                                      Alas  Babylon           2.0   
998      913                                 The Woman in Black           2.0   
999      441                  Misty of Chincoteague (Misty  #1)           2.0   

     avg_rating  
0          1.50  
1          2.00  
2          2.00  
3          2.00  
4          2.25  
..          ...  
995        5.00  
996        5.00  
997        5.00  
998        5.00  
999        5.00  

[1000 rows x 4 columns]

Variant 2

In [106]:
query_17_2 = '''
select
    books.book_id,
    books.title,
    count (distinct reviews.review_id) as review_count,
    avg (ratings.rating) as avg_rating
from books
    left join reviews on books.book_id = reviews.book_id
    left join ratings on books.book_id = ratings.book_id
group by
    books.book_id,
    books.title
order by review_count;
'''
read_sql(query_17_2)

book_id                                              title  review_count  \
0         83    Anne Rice's The Vampire Lestat: A Graphic Novel             0   
1        672   The Cat in the Hat and Other Dr. Seuss Favorites             0   
2        221                          Essential Tales and Poems             0   
3        808                            The Natural Way to Draw             0   
4        191  Disney's Beauty and the Beast (A Little Golden...             0   
..       ...                                                ...           ...   
995      733                          The Giver (The Giver  #1)             6   
996      854                                           The Road             6   
997      695  The Curious Incident of the Dog in the Night-Time             6   
998      696             The Da Vinci Code (Robert Langdon  #2)             6   
999      948                            Twilight (Twilight  #1)             7   

     avg_rating  
0      3.666667  
1      5.000000  
2      4.000000  
3      3.000000  
4      4.000000  
..          ...  
995    3.750000  
996    3.772727  
997    4.081081  
998    3.830508  
999    3.662500  

[1000 rows x 4 columns]

[Come back to Start](#100-bullet)

### A publisher, which has published the greatest number of books containing more than 50 pages

Variant 1

In [29]:
query_18 = '''
select books.publisher_id,
       publishers.publisher,
       count(books.book_id) as book_cnt
from books
join publishers on books.publisher_id = publishers.publisher_id
where num_pages > 50
group by books.publisher_id,
         publishers.publisher
order by book_cnt desc
limit 1;
'''
pd.io.sql.read_sql(query_18, con = engine)

publisher_id      publisher  book_cnt
0           212  Penguin Books        42

Variant 2

In [100]:
# as we know about duplicates for "Ballantine Books", let's check also, making lower registers
query_19 = '''
select 
       lower(publishers.publisher) as publisher,
       count(books.book_id) as book_cnt
from books
join publishers on books.publisher_id = publishers.publisher_id
where num_pages > 50
group by lower(publishers.publisher)
order by book_cnt desc
limit 1;
'''
read_sql(query_19)

publisher  book_cnt
0  penguin books        42

[Come back to Start](#100-bullet)

### An author who got the highest average user rating - to consider only the books with 50 and more ratings)

In [101]:
query_20 = '''
select sub.author_id,
       sub.author,
       avg (sub.rating_avg) as avg_rating_all_books
from
(select ratings.book_id,
        count (rating) as rating_cnt,
        avg (rating) as rating_avg,
        books.author_id,
        authors.author
from ratings
join books on ratings.book_id = books.book_id
join authors on books.author_id = authors.author_id
group by ratings.book_id, 
         books.author_id,
         authors.author
having count (rating) >= 50) as sub
group by sub.author_id,
         sub.author
order by avg (sub.rating_avg) desc
limit 1;
'''
read_sql(query_20)

author_id                      author  avg_rating_all_books
0        236  J.K. Rowling/Mary GrandPré              4.283844

[Come back to Start](#100-bullet)

### An average number of reviews from users who gave more than 50 ratings

In [102]:
query_21 = '''
select avg (sub2.reviews_count) as avg_reviews_count
from
(select count(reviews.review_id) as reviews_count
from
(select username,
       count (rating) as rating_cnt
from ratings
group by username
having count (rating) > 50) as sub1
left join reviews on sub1.username = reviews.username
group by sub1.username) as sub2;
'''
read_sql(query_21)

avg_reviews_count
0          24.333333

[Come back to Start](#100-bullet)

## Conclusions

1. There have been published 819 books since January 1, 2000.
2. For each book in the `books` dataset a number of reviews is counted (and might be from 0 till 7) and an average rating of users is displayed (might be from 1.5 till 5).
3. The publisher with the greatest number of books thicker than 50 pages: "Penguin Books" - published 42 such books.
4. The authors with the highest average book's rating (for only books with 50 or more ratings) are co-authors J.K. Rowling and Mary GrandPré - they got the average rating of 4.283844.
5. Average number of reviews from users who gave more than 50 ratings is 24.333333.

[Come back to Start](#100-bullet)